src/rgw/rgw_acl_swift.cc

```diff
@@ -59,56 +59,133 @@ static bool uid_is_public(string& uid)
+        ldout(cct, 10) << "Unknown designator: " << uid << dendl;
+	if (!result) result = -EUNKNOWN_DESIGNATOR;
+      }
+      acl.add_grant(&grant);
+    } else {
       rgw_user user(uid);
       if (rgw_get_user_info_by_uid(store, user, grant_user) < 0) {
+	// also catches ".rlistings" case - no separate mech for this in ceph
         ldout(cct, 10) << "grant user does not exist:" << uid << dendl;
-        /* skipping silently */
+        /* skipping not so silently */
+	if (!result) result = -ENO_SUCH_USER;
       } else {
         grant.set_canon(user, grant_user.display_name, perm);
         acl.add_grant(&grant);
       }
     }
   }
+  return result;
 }
 
-bool RGWAccessControlPolicy_SWIFT::create(RGWRados *store, rgw_user& id, string& name, string& read_list, string& write_list)
+int RGWAccessControlPolicy_SWIFT::create(RGWRados *store, rgw_user& id, string& name, const char* read_list, const char* write_list, uint32_t& rw_mask)
 {
   acl.create_default(id, name);
   owner.set_id(id);
   owner.set_name(name);
+  rw_mask = 0;
+  int result = 0;
 
-  if (read_list.size()) {
+  if (read_list) {
     list<string> uids;
     int r = parse_list(read_list, uids);
     if (r < 0) {
-      ldout(cct, 0) << "ERROR: parse_list returned r=" << r << dendl;
-      return false;
+      ldout(cct, 0) << "ERROR: parse_list returned r=" << r << " rl=" << read_list << dendl;
+      result = r;
     }
 
-    add_grants(store, uids, SWIFT_PERM_READ);
+    r = add_grants(store, uids, SWIFT_PERM_READ);
+    if (r) {
+      ldout(cct, 0) << "ERROR: add_grants returned r=" << r << " rl=" << read_list << dendl;
+      result = r;
+    }
+    rw_mask |= SWIFT_PERM_READ;
   }
-  if (write_list.size()) {
+  if (write_list) {
     list<string> uids;
     int r = parse_list(write_list, uids);
     if (r < 0) {
-      ldout(cct, 0) << "ERROR: parse_list returned r=" << r << dendl;
-      return false;
+      ldout(cct, 0) << "ERROR: parse_list returned r=" << r << " wl=" << write_list << dendl;
+      result = r;
     }
 
-    add_grants(store, uids, SWIFT_PERM_WRITE);
+    r = add_grants(store, uids, SWIFT_PERM_WRITE);
+    if (r) {
+      ldout(cct, 0) << "ERROR: add_grants returned r=" << r << " wl=" << write_list << dendl;
+      result = r;
+    }
+    rw_mask |= SWIFT_PERM_WRITE;
+  }
+  return result;
+}
+
+void RGWAccessControlPolicy_SWIFT::filter_merge(uint32_t rw_mask,
+                                                RGWAccessControlPolicy_SWIFT *old)
+{
+  /* rw_mask&SWIFT_PERM_READ => setting read acl,
+   * rw_mask&SWIFT_PERM_WRITE => setting write acl
+   * when bit is cleared, copy matching elements from old.
+   */
+  if (rw_mask == (SWIFT_PERM_READ|SWIFT_PERM_WRITE)) {
+    return;
+  }
+  rw_mask ^= (SWIFT_PERM_READ|SWIFT_PERM_WRITE);
+  for (auto &iter: old->acl.get_grant_map()) {
+    ACLGrant& grant = iter.second;
+    uint32_t perm = grant.get_permission().get_permissions();
+#if 0	// stuff from master, which has referrer acls
```

Comment:
would prefer not to add this stuff - if anything, maybe just a `// TODO: referrer acls`